src/gromacs/applied_forces/metatomic/metatomic_forceprovider.cpp

```diff
@@ -0,0 +1,564 @@
+        for (const auto& request_ivalue : requests_ivalue.toList())
+        {
+            nl_requests_.push_back(
+                    request_ivalue.get().toCustomClass<metatomic_torch::NeighborListOptionsHolder>());
+        }
+
+        // TODO(rg): determine device
+        model_.to(device_);
+
+        // Set data type locally on main rank
+        if (capabilities_->dtype() == "float64")
+        {
+            dtype_ = torch::kFloat64;
+        }
+        else if (capabilities_->dtype() == "float32")
+        {
+            dtype_ = torch::kFloat32;
+        }
+        else
+        {
+            GMX_THROW(APIError("Unsupported dtype from model: " + capabilities_->dtype()));
+        }
+
+        // 5. Set up evaluation options (only on main rank)
+        evaluations_options_ = torch::make_intrusive<metatomic_torch::ModelEvaluationOptionsHolder>();
+        evaluations_options_->set_length_unit("nm");
+
+        auto outputs = capabilities_->outputs();
+        if (!outputs.contains("energy"))
+        {
+            GMX_THROW(APIError("Metatomic model must provide an 'energy' output."));
+        }
+
+        auto requested_output      = torch::make_intrusive<metatomic_torch::ModelOutputHolder>();
+        requested_output->per_atom = false;
+        requested_output->explicit_gradients = {}; // Use autograd for forces
+
+        evaluations_options_->outputs.insert("energy", requested_output);
+    }
+    else
+    {
+        // non-main ranks do not load model or query capabilities, but they must still participate
+        // in collectives and must know dtype_ afterwards; dtype_ will be broadcast below.
+    }
+
+    // Initialize ML atom group data structures (must run on all ranks)
+    const auto& mtaIndices = options_.params_.mtaIndices_;
+    const int   n_atoms    = static_cast<int>(mtaIndices.size());
+
+    // Resize vectors that will be used on all ranks.
+    positions_.resize(n_atoms);
+    atomNumbers_.resize(n_atoms);
+    idxLookup_.resize(n_atoms); // Filled by gatherAtomNumbersIndices
+
+    // Atom numbers are static. We will populate them on the main rank
+    // and "broadcast" using sumReduce.
+
+    // Ensure non-main ranks have zeroed atomNumbers_ so a sumReduce acts like a broadcast.
+    std::fill(atomNumbers_.begin(), atomNumbers_.end(), 0);
+
+    // Now, only the main rank populates its version of the data.
+    if (mpiComm_.isMainRank())
+    {
+        for (int i = 0; i < n_atoms; ++i)
+        {
+            const int gIdx = mtaIndices[i];
+
+            if (gIdx >= options_.params_.numAtoms_)
+            {
+                GMX_THROW(APIError("Metatomic atom index " + std::to_string(gIdx) + " is out of bounds for topology numAtoms="
+                                   + std::to_string(options_.params_.numAtoms_)));
+            }
+            atomNumbers_[i] = options_.params_.atoms_.atom[gIdx].atomnumber;
+        }
+    }
```

Comment:
We should also error here if an atomnumber is not known by the model.

Reply:
@Luthaf said it should be done but kinda I end up in an infinite loop when I run our argon test system with current PET-MAd which does not support argon...

Reply:
Okay, it is fine we error in metatomic